### PR TITLE
Add `Counted` decorator for `Operations`. Implements #23

### PR DESCRIPTION
**Pull request by dmfs**
The `Counted` decorator sets expected number of affected rows. If the actual number of affected rows doesn't match the given number the entire transaction will fail. This can be used with Update, Delete and Assert Operations.
This also adds an `Empty` decorator which is a special case of `Counted` with an expected number of affected rows of zero.

### CODE REVIEW

File: contentpal/src/main/java/org/dmfs/android/contentpal/operations/Empty.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.dmfs.android.contentpal.TransactionContext;
+import org.dmfs.optional.Optional;
+
+
+/**
+ * An {@link Operation} which fails if the operation affects (updates, deletes or asserts) any rows at all.
+ *
+ * @author Marten Gajda
+ */
+public final class Empty<T> implements Operation<T>
```
Comment (lemonboston) [inline contentpal/src/main/java/org/dmfs/android/contentpal/operations/Empty.java]: Is the idea for this class to be used along with `Assert` for rows that should not be in the db?
Could we maybe have dedicated `Assert` classes for that, like `AssertOneRow` which would be composed of an `Assert` and `Counted(1)` and something like `AssertNoRow` which would be `Assert` and `Counted(0)`?
So I am just wondering if this is `Empty` is needed to stand on its own if this is only going to be used along with `Assert`, also because the name is not obvious for me. Or does it refer to empty table? But then again, we could have a `AssertEmpty(Table)` class for that combining `Assert` without data, and `Count(0)`.

File: contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+
+/**
+ * @author Marten Gajda
+ */
+@RunWith(RobolectricTestRunner.class)
```
Comment (lemonboston) [inline contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java]: This test runs without Roboelectirc as well.

File: contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                        }).reference(),
+                sameInstance(testReference));
+    }
+
+
+    @Test
+    public void testContentOperationBuilder() throws Exception
+    {
+        final ContentProviderOperation.Builder testBuilder = mock(ContentProviderOperation.Builder.class);
+        when(testBuilder.withExpectedCount(10)).thenReturn(testBuilder);
```
Comment (lemonboston) [inline contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java]: This test is a bit hard to follow for me, I see eventually that it tests what it has to, but I am wondering here as well whether it could look more straightforward with more Mockito usage.

**Review by lemonboston — APPROVED** [review]
Looks correct to me.

File: contentpal/src/main/java/org/dmfs/android/contentpal/operations/Empty.java (hunk 1 above)
Comment (dmfs) [reply contentpal/src/main/java/org/dmfs/android/contentpal/operations/Empty.java]: Right, I guess the only use case for this is asserts. I'll remove it.

File: contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java (hunk 3 above)
Comment (dmfs) [reply contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java]: hmm... I'll look into that.

**Comment by dmfs** [comment]
@lemonboston ready for review again.

File: contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        doReturn(testReference).when(mockOperation).reference();
+
+        assertThat(new Counted<>(1, mockOperation).reference(), sameInstance(testReference));
+    }
+
+
+    @Test
+    public void testContentOperationBuilder() throws Exception
+    {
+        final ContentProviderOperation.Builder testBuilder = mock(ContentProviderOperation.Builder.class, new FailAnswer());
```
Comment (lemonboston) [inline contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java:52]: These tests looks very good now! I didn't know this `Answer` stubbing feature, nice.
Much easier for me to get a quicker understanding of the tests now.

One totally optional thing maybe, to not re-use that `testBuilder` here but use a separate instance, that would make it even clearer for me. Like:
```
        ContentProviderOperation.Builder mockBuilder = mock(ContentProviderOperation.Builder.class, new FailAnswer());
        ContentProviderOperation.Builder resultBuilderDummy = mock(ContentProviderOperation.Builder.class, new FailAnswer());
        doReturn(resultBuilderDummy).when(mockBuilder).withExpectedCount(10);
```

Also as I see these mocking tests with Mockito can use differrent kind of 'test doubles'. I am wondering, would it make sense to start to try to use a terminology for naming the variables, based on this categorization for example?:
https://martinfowler.com/articles/mocksArentStubs.html

> **Dummy** objects are passed around but never actually used. Usually they are just used to fill parameter lists.
> **Fake** objects actually have working implementations, but usually take some shortcut which makes them not suitable for production (an in memory database is a good example).
> **Stubs** provide canned answers to calls made during the test, usually not responding at all to anything outside what's programmed in for the test.
> **Spies** are stubs that also record some information based on how they were called. One form of this might be an email service that records how many messages it was sent.
> **Mocks** are what we are talking about here: objects pre-programmed with expectations which form a specification of the calls they are expected to receive.

Probably it wouldn't always be clear, but still for example a distinction between a Mock and a Dummy - which can both be created technically as Mockito.mock() - may make things clearer in test. Or a mock() with `FailAnswer` could be viewed as a stub if it is not set up for other expectation, if it is then it becomes a Mock.
Not important of course, just a general idea/question.

**Review by lemonboston — APPROVED** [review]

File: contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java (hunk 4 above)
Comment (dmfs) [reply contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java:52]: I've deliberately chosen `doReturn(testBuilder).when(testBuilder).withExpectedCount(10);` because it models the reality in which [`withExpectedCount(10)` always returns the same instance](https://developer.android.com/reference/android/content/ContentProviderOperation.Builder.html#withExpectedCount(int))

Good Point about the names. This can make a lot of sense for more complex tests. For small tests, although still a good idea, the benefit is probably not so big.

Comment (lemonboston) [reply contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java:52]: Oh, I see the builder returns itself.. I understand it then.
